[jscoq] Port to Coq 8.13

ejgallego opened the pull request:
Nothing too remarkable, code-wise, but a big change upstream is the
addition of `zarith` as a hard dependency.

TODO:

- update addons

ejgallego commented:
It took me little to port this, however npm has a quite painful workflow and I've lost a huge amount of time, currently `make jscoq` does.
```
(cd _build/jscoq+32bit && /usr/bin/node cli.js coq-jslib/metadata/coq-pkgs.json --rootdir _vendor+v8.13+32bit/coq --nostdlib)
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module 'worker_threads'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.worker_threads (/home/egallego/research/jscoq/_build/jscoq+32bit/cli.js:26942:18)
    at __webpack_require__ (/home/egallego/research/jscoq/_build/jscoq+32bit/cli.js:26965:42)
    at Module../node_modules/fflate/esm/index.mjs (/home/egallego/research/jscoq/_build/jscoq+32bit/cli.js:2325:72)
    at __webpack_require__ (/home/egallego/research/jscoq/_build/jscoq+32bit/cli.js:26965:42)
    at Object../node_modules/fflate-unzip/lib/index.js (/home/egallego/research/jscoq/_build/jscoq+32bit/cli.js:2237:18)
    at __webpack_require__ (/home/egallego/research/jscoq/_build/jscoq+32bit/cli.js:26965:42)
```
which it is beyond my dechipering capabilities.

ejgallego commented:
Ok I found the problem indeed, due to us not supporting the stable LTS npm version.

corwin-of-amber commented:
Oh does this happen with NPM 7.x? They have introduced quite a few changes between 6.x and 7.x.

corwin-of-amber commented on Makefile:
```diff
@@ -175,7 +175,7 @@ COQ_PATCHES|Darwin/32 = byte-only
 $(COQSRC):
 	git clone --depth=1 -b $(COQ_BRANCH) $(COQ_REPOS) $@
-	cd $@ && git apply ${foreach p,$(COQ_PATCHES),$(current_dir)/etc/patches/$p.patch}
+	# cd $@ && git apply ${foreach p,$(COQ_PATCHES),$(current_dir)/etc/patches/$p.patch}
```
I assume we will have to rebuild the patches if we want to restore all functionality?

ejgallego replied:
Yes, it is in the TODO, will fix ASAP.

ejgallego replied:
Done.

corwin-of-amber commented on coq-js/dune:
```diff
@@ -21,7 +21,7 @@
  (link_flags -linkall -no-check-prims)
  ; The old makefile set: -noautolink -no-check-prims
- (libraries num jscoqlib))
+ (libraries num zarith_stubs_js jscoqlib))
```
What is `zarith_stubs_js`?

ejgallego replied:
Coq now uses `zarith`, which is an OCaml binding to `gmp` + other goodies. Given that this is a C library, we need to link to the zarith js stubs, fortunately these already exist and are maintaned.

corwin-of-amber replied:
Oh lucky us! This is also a good heads up for waCoq, will have to see how I compile zarith to WASM. (I managed to compile `num` quite easily before, but `gmp` is definitely going to be tricky; however I wanted to have a WASM port of `gmp` anyway for different reasons.)

ejgallego replied:
Yeah that looks interesting!

corwin-of-amber commented:
The PR says "merge 1 commit into `v8.12` from `v8.13`". Surely that's not what you meant.

ejgallego commented:
> The PR says "merge 1 commit into `v8.12` from `v8.13`". Surely that's not what you meant.

I just posted a PR for review, but indeed this PR should not be merged, once we set the main branch to v8.13 it should be closed?

ejgallego commented:
So pending the addons and patches this is more or less OK, SerAPI v8.13 needs some more tweaks but none of them should be very important for jscoq [famous last words]

Next thing in the roadmap IMHO is to finish some dune work that would allow seamless building of the addons, including providing to the CLI packager the right information on terms of files, coqdep output, etc... Dune already knows all these parameters so it should be not too hard.

ejgallego commented:
Ok, patches done, testing docker now; I'll let you fix addons as I'm unsure how they work yet.

I found an interesting problem, which is that upstream the format of `serapi_goals` changed, so indeed this is making jsCoq crash.

Indeed, it would be great if we could generate from the `.mli` files a corresponding TypeScript type declaration so things are checked to be in sync in both sides, we have a tool that does the dual https://github.com/LexiFi/gen_js_api , but I think nothing on the other direction yet.

ejgallego commented:
For the goals problem, I've fixed serapi upstream as to recover those fields again.

ejgallego commented:
> Oh does this happen with NPM 7.x? They have introduced quite a few changes between 6.x and 7.x.

This happens with npm 6, need to pass the --experimental-worker option to it.

ejgallego commented:
Ok @corwin-of-amber I'm satisfied with this so far; once you give the green light we can make this branch the default.

corwin-of-amber commented:
> Indeed, it would be great if we could generate from the `.mli` files a corresponding TypeScript type declaration so things are checked to be in sync in both sides, we have a tool that does the dual https://github.com/LexiFi/gen_js_api , but I think nothing on the other direction yet.

Intriguing!

About the docker: I originally used `./toolchain-setup.sh --64` (64-bit) so that the same Dockerfile could work on my macOS. Since then I changed `./toolchain-setup.sh` so that `--64` is the default on macOS, so I recommend to remove the flag from the Dockerfile so that the default is used.
(I am not sure what the situation is when a Linux Docker container is run inside macOS; at any rate, I am not running Docker on my personal laptop right now since I no longer have an Intel processor, I am only running it on a Linux host anyway.)

corwin-of-amber commented:
I would have a closer look at the error with npm 6. Perhaps the error was caused because I checked in a `package-lock.json` that was generated using the newer npm? In our specific case I believe it is not crucial to save `package-lock.json` if it breaks compatibility (although it is good practice in general).

corwin-of-amber commented:
I got this error building `coq-pkgs`:
```
cannot open workspace 'coq-jslib/metadata/coq-pkgs.json': Error: ENOENT: no such file or directory, scandir '_vendor+v8.13+64bit/coq/plugins/setoid_ring'
```
The CLI just skips this directory, but it hints that the plugins directory structure has changed. Better see if `coq-pkgs.json` is coherent.

ejgallego commented:
> About the docker: I originally used `./toolchain-setup.sh --64` (64-bit) so that the same Dockerfile could work on my macOS. Since then I changed `./toolchain-setup.sh` so that `--64` is the default on macOS, so I recommend to remove the flag from the Dockerfile so that the default is used.

IMHO we should try to make Docker images the most reproducible possible, so I'd rather add more parameters than less. I'm surprised there was some issue under OSX with --32, likely the problem was apt [which can be fixed, but in a different PR]

> At any rate, I see no reason not to make v8.13 the default branch. I have started a feature branch (locally) vs. v8.12 but that branch is not going anywhere so I will rebase it eventually.

Great! I'll do once now as I'm satisfied with my testing. Then we can do as we do in Coq, backport critical fixes to 8.12.

> I would have a closer look at the error with npm 6. Perhaps the error was caused because I checked in a `package-lock.json` that was generated using the newer npm? In our specific case I believe it is not crucial to save `package-lock.json` if it breaks compatibility (although it is good practice in general).

It was due to several issues it seems; I am OK upgrading to npm to 7, but we should say so, and add some kind of sanity check.

I also saw some issue with the versioning, due to the -rc version set differently in `package.json` vs `package.lock.json`

> I got this error building coq-pkgs:

The ring plugin was renamed, fixed [note, only the ml files]

ejgallego commented:
Default branch changed.

corwin-of-amber commented:
> The ring plugin was renamed, fixed [note, only the ml files]

Also, `theories/Array` was apparently added — should we add it to `coq-collections`?

ejgallego commented:
Indeed. Done, with some tweaks as these modules depend on arith.

corwin-of-amber commented on docs/build.md:
```diff
@@ -36,14 +36,14 @@ cd jscoq
 For Dune builds, configure the switch in your `dune-workspace`.
 
- 3. Fetch Coq 8.10 sources from the repository and configure it for build.
+ 3. Fetch Coq 8.13 sources from the repository and configure it for build.
```
Ha good catch 😄
